### PR TITLE
bazel: similar build flags to cmake for corresponding build types (#4…

### PR DESCRIPTION
**Pull request by htuch**
…15).

Bazel has 3 compilation modes out-of-the-box {fastbuild, opt, dbg}, see
https://bazel.build/versions/master/docs/bazel-user-manual.html#flag--compilation_mode.

The cmake build flags for normal/debug/server_only/asan/coverage are at
https://github.com/lyft/envoy/issues/415#issuecomment-292592703.

Below are the relevant flags (ignoring -include etc.) from the various builds. Lines that begin with
* are those that include additional flags we're injecting.

fastbuild:

 -U_FORTIFY_SOURCE -fstack-protector -Wall
 -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x'
 -frandom-seed=...
 -fPIC
*-DBAZEL_BRINGUP -Wall -Wextra -Werror -Wnon-virtual-dtor -Woverloaded-virtual
*-Wold-style-cast '-std=c++0x' -includeprecompiled/precompiled.h -DDEBUG
 -fno-canonical-system-headers
 -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"'
 '-D__TIME__="redacted"'

dbg:

 -U_FORTIFY_SOURCE -fstack-protector -Wall
 -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g '-std=c++0x'
 -frandom-seed=...
 -fPIC
*-DBAZEL_BRINGUP -Wall -Wextra -Werror -Wnon-virtual-dtor
*-Woverloaded-virtual -Wold-style-cast '-std=c++0x' -includeprecompiled/precompiled.h -ggdb3 -DDEBUG
 -fno-canonical-system-headers
 -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"'
 '-D__TIME__="redacted"'

opt:

 -U_FORTIFY_SOURCE -fstack-protector -Wall
 -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2
 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x'
 -frandom-seed=...
*-DBAZEL_BRINGUP -Wall -Wextra -Werror -Wnon-virtual-dtor
*-Woverloaded-virtual -Wold-style-cast '-std=c++0x' -includeprecompiled/precompiled.h
 -fno-canonical-system-headers
 -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"'
 '-D__TIME__="redacted"'

This PR requires #716 to merge, since #716 provides some important changes to how the headers are
exported from the 3rd party deps to allow -Wextra and -Wold-style-cast to work.

I've verified that we now have a 100% static envoy-static build, which ldd tells me is not a dynamic
binary.

% bazel build -c opt //source/exe:envoy-static.stripped
Target //source/exe:envoy-static.stripped up-to-date:
  bazel-bin/source/exe/envoy-static.stripped

% ldd bazel-bin/source/exe/envoy-static.stripped
        not a dynamic executable

% ls -lh bazel-bin/source/exe/envoy-static.stripped
 -r-xr-xr-x 1 htuch eng 8.1M Apr  7 15:46 bazel-bin/source/exe/envoy-static.stripped

### CODE REVIEW

**Comment by mattklein123** [comment]
>  ldd bazel-bin/source/exe/envoy-static.stripped
not a dynamic executable

That seems not right, or at least not what I'm used to seeing.

```
mklein@localhost:~/Source/envoy-private$ ldd build_local/envoy/source/exe/envoy
	linux-vdso.so.1 =>  (0x00007ffe86975000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f81f5b05000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f81f58fd000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f81f55f3000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f81f522a000)
	/lib64/ld-linux-x86-64.so.2 (0x00005582f81a7000)
```

Are you somehow statically linking everything? I've had a lot of issues with full static linking and would prefer we still dynamically link to system libraries (or at least have that option).

**Comment by mattklein123** [comment]
I dropped a note into the build GH issue, but FYI we also need to link in build ID. (This change doesn't need to cover that but since we were discussing parity here I thought i would mention it).

**Comment by htuch** [comment]
Waiting for input on https://github.com/bazelbuild/bazel/issues/2805 for the `--build-id` link issue.

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -3,22 +3,19 @@ load("@protobuf_git//:protobuf.bzl", "cc_proto_library")
 ENVOY_COPTS = [
     # TODO(htuch): Remove this when Bazel bringup is done.
     "-DBAZEL_BRINGUP",
-    "-fno-omit-frame-pointer",
```
Comment (mattklein123) [inline bazel/envoy_build_system.bzl]: Is this default in bazel?

File: bazel/envoy_build_system.bzl [hunk 2]
```diff
@@ -3,22 +3,19 @@ load("@protobuf_git//:protobuf.bzl", "cc_proto_library")
...
     "-Woverloaded-virtual",
-    # TODO(htuch): Figure out how to use this in the presence of headers in
-    # openssl/tclap which use old style casts.
-    # "-Wold-style-cast",
+    "-Wold-style-cast",
     "-std=c++0x",
     "-includeprecompiled/precompiled.h",
-]
+] + select({
+    "//bazel:opt_build": [],
```
Comment (mattklein123) [inline bazel/envoy_build_system.bzl]: -DNEDEBUG? Still need debug symbols?

Comment (htuch) [inline bazel/envoy_build_system.bzl]: Bazel automatically sets `NDEBUG` in `opt`, see the output in the above commit messages. We're now actually in a strange situation where Bazel is setting `NDEBUG` in opt and we set `DEBUG` in fastbuild/dbg. Shouldn't we just have one define indicating debug and use the negated logic where necessary?

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Yep.

**Comment by htuch** [comment]
The `--build-id` fix is in https://github.com/lyft/envoy/pull/767, so we can ship this one once it's approved and tests pass.

File: bazel/envoy_build_system.bzl (hunk 2 above)
Comment (mattklein123) [inline bazel/envoy_build_system.bzl]: I think DEBUG is just habit from previous jobs. I don't think we use it anywhere and I agree it's silly. Let's just kill it. Can you grep for it though just to make sure we don't use it anywhere?

**Comment by mattklein123** [comment]
What's preventing this from passing tests? build image upgrade?

File: bazel/envoy_build_system.bzl (hunk 2 above)
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Confirmed it's not used, switching to NDEBUG only.

**Comment by htuch** [comment]
It wasn't building as the headers included from external deps were subject to the `-Wold-style-cast -Wextra` treatment. I've changed how they get included so that they are now treated as "system" headers, by using `includes` rather than `strip_include_prefix` to reference them in `ci/prebuild/BUILD`. Will push after validating on all Bazel flows and build types.

**Pull request merged**